_posts/2024-04-28-changing-loggers-level-from-unleash.adoc

```diff
@@ -0,0 +1,562 @@
+            }
+        }
+    }
+
+    private LogConfig[] getLogConfigs() {
+        Variant variant = unleash.getVariant(UNLEASH_TOGGLE_NAME); // <3>
+        if (variant.isEnabled()) { // <4>
+            Optional<Payload> payload = variant.getPayload();
+            if (payload.isPresent() && payload.get().getType().equals("json") && payload.get().getValue() != null) {
+                try {
+                    return objectMapper.readValue(payload.get().getValue(), LogConfig[].class);
+                } catch (JsonProcessingException e) {
+                    Log.error("Variant payload deserialization failed", e);
+                }
+            }
+        }
+        return new LogConfig[0]; // <5>
+    }
+
+    private void setLoggerLevel(String loggerName, String levelName) {
+        Logger logger = Logger.getLogger(loggerName);
+        Level currentLevel = logger.getLevel();
+        Level newLevel = Level.parse(levelName);
+        if (!newLevel.equals(currentLevel)) {
+            logger.setLevel(newLevel);
+        }
+    }
+}
+----
+<1> From now on, `LogLevelManager` is an `@ApplicationScoped` bean.
+<2> `Unleash` is an `@ApplicationScoped` bean produced by the https://docs.quarkiverse.io/quarkus-unleash/dev/index.html[quarkus-unleash^] extension.
+<3> Be careful about the argument passed to `Unleash#getVariant`: it has to be the toggle name, not the variant name.
+<4> `variant.isEnabled()` will return `false` if the toggle is disabled in Unleash or if the toggle has no variants.
+<5> If the method is unable to find a variant payload or if it fails to deserialize that payload for any reasons, an empty `LogConfig` array will be returned.
+
+We can now retrieve the loggers configuration from Unleash, that's great!
+But that new `LogLevelManager#updateLoggerslevel` method isn't used yet.
+Where should it be used from, and when?
+
+image::triggering.png[Triggering the loggers level update]
+
+We need that method to be executed as soon as the loggers configuration is changed in Unleash.
+So, its execution has to be periodically scheduled somehow.
+We could make the method `@Scheduled` with the https://quarkus.io/guides/scheduler-reference[quarkus-scheduler^] extension, but there is a better approach thanks to the Unleash SDK.
+Let's jump to the next section.
+
+=== The Subscriber API from Unleash
+
+The Unleash Client SDK for Java comes with a feature that will be very helpful here: the https://docs.getunleash.io/reference/sdks/java#subscriber-api[Subscriber API^].
+The https://github.com/Unleash/unleash-client-java/blob/main/src/main/java/io/getunleash/event/UnleashSubscriber.java[UnleashSubscriber^] interface can indeed by implemented to subscribe to various Unleash events, including `FeatureToggleResponse` which is emitted when the Unleash client fetches toggles from the server.
+
+Using the Subscriber API with the https://docs.quarkiverse.io/quarkus-unleash/dev/index.html[quarkus-unleash^] extension is extremely simple.
```

Comment:
I'm curious: why the `^` at the end of the link text?

Reply:
I'm an Asciidoc beginner so it may be wrong, but I added that to open the link in a new browser tab.

Reply:
Ah okay. In general, I recommend to let the users decide what they want to do. But OK, I wasn't aware of this AsciiDoc feature.